indexing-service/src/main/java/org/apache/druid/indexing/overlord/ImmutableWorkerInfo.java

```diff
@@ -119,12 +142,33 @@ public boolean isValidVersion(String minVersion)
     return worker.getVersion().compareTo(minVersion) >= 0;
   }
 
-  public boolean canRunTask(Task task)
+  public boolean canRunTask(Task task, double parallelIndexWorkerRatio)
   {
     return (worker.getCapacity() - getCurrCapacityUsed() >= task.getTaskResource().getRequiredCapacity()
+            && canRunParallelIndexTask(task, parallelIndexWorkerRatio)
             && !getAvailabilityGroups().contains(task.getTaskResource().getAvailabilityGroup()));
   }
 
+  public boolean canRunParallelIndexTask(Task task, double parallelIndexWorkerRatio)
+  {
+    if (!task.getType().equals(ParallelIndexSupervisorTask.TYPE)) {
+      return true;
+    }
+    return getWorkerParallelIndexCapacity(parallelIndexWorkerRatio) - getCurrParallelIndexCapacityUsed()
```

Comment:
Will this strategy result in artificially reducing parallel task supervisor concurrency? For example, assume that you have 10 slots and parallelIndexWorkerRatio is 0.1 . Assume that one ingestion is running with one supervisor an one parallel sub task running then if another parallel ingestion is started even though there are 8 slots the new parallel task won't run because the parallel task slots (1) has already been consumed..

Reply:
Yes, that could be an issue with the default of 0.1

We could set the default value to 1, in which case the behaviour would be the same as it is today.

A default value of 0.5 might be the best? Deadlocks won't occur and subtask concurrency is not affected except in the cases where there a lot of supervising tasks are running

The value can be configured to be lower when progress supervising tasks are configured to have a lot of subtasks and "fast" progress of a few ingestion tasks is preferred over "slow" progress of all the submitted ones

Reply:
Did you look into considering actually have the supervisor task not take any slots? In this way the limiting factor would be the number of sub-tasks allowed but all tasks now are allowed to run with no starvation of any of them (since their supervisor tasks are always allowed to run and their sub-tasks will eventually run when slots free up, assuming that there is already some fairness built in to guarantee older waiting sub-tasks get to run before newer or something like that -- I am not sure that this exists). The potential issue with letting all supervisor run with no task slots taken is that while they don't use a lot of cpu they may use significant memory given that they keep track of the metadata of running sub-tasks, but this can be deal with by adding more memory to the middle managers in case of OOMs.